src/java/org/apache/cassandra/io/sstable/format/bti/BtiTableReaderLoadingBuilder.java

```diff
@@ -0,0 +1,241 @@
+                builder.setLast(partitioner.decorateKey(builder.getStatsMetadata().lastKey));
+            }
+
+            if (builder.getComponents().contains(Components.PARTITION_INDEX))
+            {
+                builder.setPartitionIndex(openPartitionIndex(builder.getFilter().isInformative()));
+                if (builder.getFirst() == null || builder.getLast() == null)
+                {
+                    builder.setFirst(builder.getPartitionIndex().firstKey());
+                    builder.setLast(builder.getPartitionIndex().lastKey());
+                }
+            }
+
+            try (CompressionMetadata compressionMetadata = CompressionInfoComponent.maybeLoad(descriptor, components))
+            {
+                builder.setDataFile(dataFileBuilder(builder.getStatsMetadata()).withCompressionMetadata(compressionMetadata).complete());
+            }
+        }
+        catch (IOException | RuntimeException | Error ex)
+        {
+            // in case of failure, close only those components which have been opened in this try-catch block
+            Throwables.closeAndAddSuppressed(ex, builder.getPartitionIndex(), builder.getRowIndexFile(), builder.getDataFile(), builder.getFilter());
+            throw ex;
+        }
+    }
+
+    private IFilter buildBloomFilter(StatsMetadata statsMetadata) throws IOException
+    {
+        IFilter bf = null;
+
+        try (KeyReader keyReader = createKeyReader(statsMetadata))
+        {
+            bf = FilterFactory.getFilter(statsMetadata.totalRows, tableMetadataRef.getLocal().params.bloomFilterFpChance);
+
+            while (!keyReader.isExhausted())
+            {
+                DecoratedKey key = tableMetadataRef.getLocal().partitioner.decorateKey(keyReader.key());
+                bf.add(key);
+
+                keyReader.advance();
+            }
+        }
+        catch (IOException | RuntimeException | Error ex)
+        {
+            Throwables.closeAndAddSuppressed(ex, bf);
+            throw ex;
+        }
+
+        return bf;
+    }
+
+    private IFilter loadFilter(ValidationMetadata validationMetadata)
+    {
+        return FilterComponent.maybeLoadBloomFilter(descriptor,
+                                                    components,
+                                                    tableMetadataRef.get(),
+                                                    validationMetadata);
+    }
+
+    private PartitionIndex openPartitionIndex(boolean preload) throws IOException
+    {
+        try (FileHandle indexFile = partitionIndexFileBuilder().complete())
+        {
+            return PartitionIndex.load(indexFile, tableMetadataRef.getLocal().partitioner, preload);
+        }
+        catch (IOException ex)
+        {
+            logger.debug("Partition index file is corrupted: " + descriptor.fileFor(Components.PARTITION_INDEX), ex);
+            throw ex;
+        }
+    }
+
+    private FileHandle.Builder dataFileBuilder(StatsMetadata statsMetadata)
```

Comment:
This method looks duplicated exactly in `BigSSTableReaderLoadingBuilder`, so I'm guessing we could move it into `SSTableReaderLoadingBuilder` (along w/ the `dataFileBuilder` field, which isn't `final`).